usr/iscsid.c

```diff
@@ -272,25 +275,45 @@ static int sync_session(__attribute__((unused))void *data,
 	/* multiple drivers could be connected to the same portal */
 	if (!iscsi_match_session(&rec, info))
 		return -1;
+
+	req = calloc(1, sizeof(*req));
+	if (!req)
+		return 0;
+	INIT_LIST_HEAD(&req->list);
+
 	/*
 	 * We use the initiator name from sysfs because
 	 * the session could have come from our db or ibft or some other
 	 * app.
 	 */
 	strcpy(rec.iface.iname, info->iface.iname);
-	memset(&req, 0, sizeof(req));
-	req.command = MGMT_IPC_SESSION_SYNC;
-	req.u.session.sid = info->sid;
-	memcpy(&req.u.session.rec, &rec, sizeof(node_rec_t));
+	req->req.command = MGMT_IPC_SESSION_SYNC;
+	req->req.u.session.sid = info->sid;
+	memcpy(&req->req.u.session.rec, &rec, sizeof(node_rec_t));
 
+	list_add_tail(&req->list, sync_list);
+	return 0;
+}
+
+static void sync_sessions(struct list_head *sync_list)
+{
+	struct iscsid_async_req *req, *tmp;
+	iscsiadm_rsp_t rsp;
+	int retries = 0, rc;
+
+	list_for_each_entry_safe(req, tmp, sync_list, list) {
```

Comment:
Should you not reset retries to zero for each entry in the list, rather than having 30 retries total for all items on the list, no matter how many?

Reply:
We don't need to reset it. We actually might not even need it. I noticed the following while working on that code.

Originally, we were only handling a race where the fist request could be sent from this process, but main() was still doing the operations after the sync session fork, but had not yet done mgmt_ipc_listen and opened the iscsid socket and started listening on it.

However, I don't think that race will happen. Either I messed up and that race could never happen or someone fixed the code.

Right now, I think main() will do:

1. mgmt_ipc_listen(). That will then setup iscsid to start to listen on the iscsid socket.
2. We then fork the sync session process.
3. The sync session process might send a request.
4. If 3 happens before main does event_loop(), I think we will be ok because we already started listening on the socket in #1.

Note that for iscsistart, I think we can still race because we do the fork before mgmt_ipc_listen. I think we can change iscsistart so that it works like iscisd and then delete all that ISCSI_ERR_ISCSID_NOTCONN retry code in login_session.

I was thinking I'll wait on that GH issue about systemd and socket handling to shake out then modify this code accordingly if we some changes for that.